### PR TITLE
draft addition of aria-valuetext on more things

### LINKED ISSUE #711: aria-valuetext would be useful on many more roles
There are some places in the APG where aria-labelledby is being used to provide values. I would like to see the ability to support aria-valuetext on more roles to provide a readable value-like attribute for the item.

An example is https://www.w3.org/TR/wai-aria-practices-1.1/examples/listbox/listbox-collapsible.html Where the button has <button aria-haspopup="listbox" aria-labelledby="exp_elem exp_button" id="exp_button">Neptunium</button>

Another place where this would be essential in order to be able to mimic HTML functionality would be creating a labelled textbox. As the spec currently is written there is no way to provide both an accessible name and a value for something with role="textbox" for example.

### PR DESCRIPTION
**Pull request by jnurthen**
Very draft PR to close #711

***
<a href="https://pr-preview.s3.amazonaws.com/w3c/aria/pull/1413.html" title="Last updated on Mar 2, 2021, 1:30 AM UTC (808afa9)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/w3c/aria/1413/8c9d7b6...808afa9.html" title="Last updated on Mar 2, 2021, 1:30 AM UTC (808afa9)">Diff</a>

### CODE REVIEW

**Comment by stes-acc** [comment]
related: [role switch should have additional attribute aria-valuetext · Issue #574](https://github.com/w3c/aria/issues/574)

File: index.html [hunk 1]
```diff
@@ -4655,6 +4656,7 @@ <h2>Definition of Roles</h2>
 								<li><pref>aria-multiselectable</pref></li>
 								<li><pref>aria-readonly</pref></li>
 								<li><pref>aria-required</pref></li>
+								<li><pref>aria-valuetext</pref></li>
```
Comment (mcking65) [inline index.html:4659]: This idea makes sense for the collabsible listbox where there may be an element representing a listbox that does not have any options. However, I think we should deprecate that pattern in APG. Select-only combobox completely replaces it; the combobox is simpler to code and supports aria-required. In any other listbox, the selected option represents the value.

File: index.html [hunk 2]
```diff
@@ -9317,6 +9321,7 @@ <h5>Presentational Roles Conflict Resolution</h5>
 			<rdef>textbox</rdef>
 			<div class="role-description">
 				<p>A type of input that allows free-form text as its value.</p>
+				<p>If the <pref>aria-valuetext</pref> <a>attribute</a> is set it indicates the <a>value</a> of the textbox. If it is not set then the value is the child content of the textbox.</p>
```
Comment (mcking65) [inline index.html:9324]: I don't think this should be permitted if the textbox role is on an html input element. It needs language to the effect:

>If the textbox element is a host language element that provides a value, such as an HTML input  element, authors must not specify aria-valuetext.
>If the textbox element is a host language element that provides a value, such as an HTML input  element, user agents MuST ignore aria-valuetext if it is specified.

File: index.html [hunk 3]
```diff
@@ -13286,11 +13292,22 @@ <h2>Definitions of States and Properties (all aria-* attributes)</h2>
...
+					<li>Authors SHOULD only set the <pref>aria-valuetext</pref> attribute when the rendered value cannot be meaningfully represented as a number. For example, a slider may have rendered values of <code>small</code>, <code>medium</code>, and <code>large</code>. In this case, the values of <pref>aria-valuenow</pref> could range from 1 through 3, which indicate the position of each value in the value space, but the <pref>aria-valuetext</pref> would be one of the strings: <code>small</code>, <code>medium</code>, or <code>large</code>. If the <pref>aria-valuetext</pref> attribute is absent, the <a>assistive technologies</a> will rely solely on the <pref>aria-valuenow</pref> attribute for the current value.</li>
+					<li>If <pref>aria-valuetext</pref> is specified, assistive technologies SHOULD render that value instead of the value of <pref>aria-valuenow</pref>.</li>
+				</ul>
+				<p>When this property is used on <rref>textbox</rref>, <rref>combobox</rref> or <rref>listbox</rref>:</p>
```
Comment (mcking65) [inline index.html:13302]: Per above, I don't think we should include listbox. Also, not sure if it should be supported on textbox unless it is readonly and only if not backed by a native element that does not have its own value.

**Review by mcking65 — COMMENTED** [review]
I think this is a great start. I think the application to switch is most interesting. It is good for combobox as long as it is not editable, i.e., only on select-only comboboxes. I am fairly certain that we should not add support for listbox. And, I think textbox support needs additional restrictions. Details below.

**Review by carmacleod — COMMENTED** [review]
I'm good with switch.

I wonder if prose similar to the UA instructions we added for combobox (pasted below) would go a long way towards helping authors understand where the value for textbox and searchbox usually comes from. UAs already do this for `div role="textbox"` or `div role="searchbox"` anyhow, so we're not saying anything new.

> User agents MUST expose the value of elements with role combobox to assistive technologies. The value of a combobox is represented by one of the following:
>
> - If the combobox element is a host language element that provides a value, such as an HTML input element, the value of the combobox is the value of that element.
> - Otherwise, the value of the combobox is represented by its descendant elements and can be determined using the same method used to compute the name of a button from its descendant content.

We could maybe add another bullet point to the effect of aria-valuetext only being useful if it's (a) not a host language element that already provides a value, and (b) if the value doesn't match the descendant content.

Regarding listbox, I agree with Matt that it probably shouldn't support aria-valuetext. The selected option(s) represent the listbox value, and ATs look for that. We could maybe make this explicit in the prose. Note that UAs don't set the actual value of the listbox. It's implied by the selected option(s).

Regarding CSS content, UAs should always be including CSS text content as specified in the name calculation in https://w3c.github.io/accname/#step2F. So, "the value doesn't match the CSS content" shouldn't be a justification for adding aria-valuetext support.

**Comment by pkra** [comment]
Is 1.3 still realistic for this one?

**Pull request closed without merging**